### PR TITLE
Fix skip range miscalculation

### PR DESCRIPTION
**Pull request by matthewhughes934**
As well as a bit of code tidying

- Idempotent test: avoid redundant file reads

    We read the 'significant comments' in `check_files`, so just pass it
    down rather than re-reading the test file.

- Fix skip range miscalculated on nested blocks

    Fix a bug in `FmtVisitor::push_skipped_with_span` that assumed
    `line_number` was always absolute (within a given file) and instead set
    the end of the range to be the end of the span with the skip attribute.

    Found while investigating issue #6954 (but this commit doesn't address
    that issue).

### CODE REVIEW

File: src/visitor.rs [hunk 1]
```diff
@@ -808,7 +808,7 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
         // So here we need to take a minimum between the two.
         let lo = std::cmp::min(attrs_end + 1, first_line);
         self.push_rewrite_inner(item_span, None);
-        let hi = self.line_number + 1;
+        let hi = self.psess.line_of_byte_pos(item_span.hi());
         self.skipped_range.borrow_mut().push((lo, hi));
     }
```
Comment (ytmimi) [inline src/visitor.rs:813]: I've wondered for a while if using the spans from the **_source_** makes sense  when it comes to determining the `skipped_range`. Wouldn't we have to figure out the range relative to what's already been formatted in the file?

For example:

```rust
fn main() {
   let some.long().method().call().chain().that().will().wrap();

   #[rustfmt::skip] // line 4 before formatting;
   let x = 100;
}
```

```rust
fn main() {
   let some
        .long()
        .method()
        .call()
        .chain()
        .that()
        .will()
        .wrap();

   #[rustfmt::skip] // line 11 after formatting;
   let x = 100;
}
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/7049/1191d91d83f5582766404b4e912d19cd89866d56..7304e12e4a599db99e21ab4539b9586d8a07986d)*

File: tests/target/skip_with_overflow.rs
Comment (ytmimi) [inline tests/target/skip_with_overflow.rs:1]: Might make more sense to turn this into a snapshot test and show that stderr is empty. Alternatively you can write a [unit test like this one](https://github.com/rust-lang/rustfmt/blob/bbfd1201f7b9c7eddb7fb7dec704fa3566b6863c/tests/rustfmt/main.rs#L323-L333).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustfmt/7049/1191d91d83f5582766404b4e912d19cd89866d56..7304e12e4a599db99e21ab4539b9586d8a07986d)*

File: src/visitor.rs (hunk 1 above)
Comment (matthewhughes934) [reply src/visitor.rs:813]: > Wouldn't we have to figure out the range relative to what's already been formatted in the file?

yes :upside_down_face:, I can break my test with your change: on the first run it adds a 'skip' range that ends before the long string _after_ formatting the method chain (no errors on subsequent runs, once the chain is formatted)

```diff
diff --git i/tests/target/skip_with_overflow.rs w/tests/target/skip_with_overflow.rs
index a3a03b80..89cd7a3e 100644
--- i/tests/target/skip_with_overflow.rs
+++ w/tests/target/skip_with_overflow.rs
@@ -12,8 +12,10 @@ fn foo() {
         // the bug: we'd mark the region (lo=16, hi=10) as skipped
         // the lo value is correct, but the hi value is the offset of the line after the end of the
         // 'if' block relative to the start of the '|| {' block
-        #[rustfmt::skip]
         if true {
+            some.very().very().very().very().very().very().long().method().call().chain().that().will().wrap();
+
+            #[rustfmt::skip]
             println!(
                 "this is a very long string, it goes over max_width. This is just padding to push it over"
             );
```

Comment (matthewhughes934) [reply src/visitor.rs:813]: which leaves me with the question: do we fix the bug in the current approach (writing skips w.r.t unformatted source). Or, do we try and move to some more robust approach? I have no idea how the latter would look, some ideas that come to mind (I have not thought about them much, neither of them feel very good):

* checking for overflows on each line we push (where we have the context of the current span we're working on)
* Append some marker comment like `// _RUSTFMT_SKIP_HERE` to all lines in a skip block, so when looking at the formatted source we know what's skip (then strip out those comments before writing things out)

Comment (ytmimi) [reply src/visitor.rs:813]: A proper fix for this would definitely be more involved and it's not easy with the current design. I think we'll need to put some more thought into this one unfortunately.

**Comment by ytmimi** [comment]
Marking this as `on-hold` until we come up with a proper design.

**Comment by matthewhughes934** [comment]
Closing, this needs a different approach that needs some extra thought. While poking around I discovered another issue https://github.com/rust-lang/rustfmt/issues/7076 which needs to also be addressed when solving this issue.
